### ISSUE TITLE
Inconsistent behavior of hessian

### ISSUE THREAD

**Issue by gbaydin** [DiffSharp/DiffSharp#328]
I noticed the following is failing with an error

```fsharp
let f (x:Tensor) = x.sum()
let x = dsharp.tensor([0.5, 0.5])
dsharp.hessian f x
```

while the following is running fine

```fsharp
let f (x:Tensor) = x.exp().sum()
let x = dsharp.tensor([0.5, 0.5])
dsharp.hessian f x
```

We expect these two functions to behave the same except the resulting numerical values. I suspect this has to do with the reverse mode of `sum` but haven't checked yet.

**Comment by dsyme** [comment]
We are getting a TensorC flowing into `reversePush`.  It boils down to this:

```fsharp


open DiffSharp
open DiffSharp.Util
open DiffSharp.Numerical

let x = dsharp.tensor([0.5, 0.5])
let hv = 
    let v = dsharp.tensor([1, 0])
    let x = x |> dsharp.reverseDiff 0u
    printfn "x = %A" x
    printfn "parents(x) = %A" (x.parents())
    let a = (x.forwardDiff(v, 1u)).sum()
    printfn "parents(a) = %A" (a.parents())
    let fx = a.primal // TensorR(...).sum()
    let gv = a.derivative 
    printfn "parents(gv) = %A" (gv.parents())
    printfn "gv  = %A" gv
    gv.reverse(gv.onesLike())
    x.derivative
hv
```
Here `a` is a TensorF, `a.primal` is TensorR but `gv` = `a.derivative`  is TensorC (= `v.sum()`).   The TensorC is pushed into `reversePush`.   

This is either an error condition or represents an input independent of the reverse push and thus contributes zero to the reverse push.  If the latter then nothing ever adjsuts the scalar used for empty `zero` initial adjoint.  

If it's an error condition then my guess is that we have to ensure an invariant e.g. that if `a.primal` is TensorR then `a.derivative` is also TensorR of the same level.  

Note by simple reduction of the F# code the above currently boils down to just this:
```fsharp
open DiffSharp
open DiffSharp.Util
open DiffSharp.Numerical

let x0 = dsharp.tensor([0.5, 0.5])
let x = x0 |> dsharp.reverseDiff 0u
let gv = dsharp.tensor(1)
gv.reverse(gv.onesLike())
x.derivative
```



This arises naturally from:
//          let inline dfFwd(ap,ad:Tensor,fp) = ad.sum(?dtype=dtype)
// because ap is TensorR but ad is TensorC

// So 
//    let x0 = dsharp.tensor([0.5, 0.5])
//    let x = x0 |> dsharp.reverseDiff 0u
//    let a = (x.forwardDiff(v, 1u)).sum().derivative
// is TensorC

**Comment by dsyme** [comment]
DiffSharp today has the invariant that

    (f (x.forwardDiff(v))).primal.isReverseDiff() = (f x).isReverseDiff()

That is the primal of the forwardDiff has the same levelling/TensorR/TensorF etc as the result of applying the function without forwardDiff - the forwardDiff adds a layer.  It possible that should apply to .derivative too, so

    (f (x.forwardDiff(v))).derivative.isReverseDiff() = (f x).isReverseDiff() 

should also always be true - we can't just return the `TensorC(1)` as the derivative, but rather need to return `TensorR(TensorC(1), NewT)`.   This invariant would fix this bug, and potentially others.

For example: 
```
open DiffSharp
open DiffSharp.Util
let f (x: Tensor) = x.sum()
let x = dsharp.tensor(1).reverseDiff()
let v = dsharp.tensor(1)

open DiffSharp
open DiffSharp.Util
let f (x: Tensor) = x.sum()
let x = dsharp.tensor(1).reverseDiff()
let v = dsharp.tensor(1)
(f (x.forwardDiff(v))).primal.isReverseDiff() = (f x).isReverseDiff() |> print
(f (x.forwardDiff(v))).derivative.isReverseDiff() = (f x).isReverseDiff() |> print
```

prints "true" and "false".  If you change "f" to say `f(x) = x.exp().log().sum()` then it prints "true" and "true".

If you want this invariant, then I think this line in OpUnary:

        | TensorF(ap,ad,at)    -> let fp = fTensor(ap) in TensorF(fp, dfFwd(ap,ad,fp), at)

would need to be something like:

        | TensorF(ap,ad,at)    ->
            let fp = fTensor(ap)
            let df = dfFwd(ap,ad,fp)
            let df = .... uplevel df until it looks like fp ..... (e.g. adding TensorR(..., NewT, ..), TensorF etc.)
            TensorF(fp, df, at)

**Comment by dsyme** [comment]
Alternatively the fix could be here?

```fsharp
    member t.forwardDiff(derivative:Tensor, ?tag:uint32) = 
         ...
        let derivative = ... uplevel df until it is the same level and differentiation kind as t ...
        TensorF(t, derivative, tag)
```

**Comment by dsyme** [comment]
Note applying reverse gradient to `v` works, I'm not sure this is correct 
```fsharp
open DiffSharp
open DiffSharp.Util
open DiffSharp.Numerical

let x = dsharp.tensor([0.5, 0.5])
let v = dsharp.tensor([1, 0])                     // TensorC([1,0])
let vrev = v |> dsharp.reverseDiff 0u               // TensorR(TensorC([1,0]), ...)
let xrev = x |> dsharp.reverseDiff 0u         // TensorR([0.5,0.5], ...)
let xrevfwd = xrev.forwardDiff(v, 1u)         // TensorF(TensorR([0.5,0.5], ...), TensorR(TensorC([1,0]),...), ...)
let a = xrevfwd.sum()                         // TensorF(TensorR(1, SumT, ...), TensorR(TensorC(1), SumT, ...))
let gv = a.derivative                         // TensorR(TensorC(1), SumT, ...)
gv.reverse(gv.onesLike())                     //  push 1
xrev.derivative                               //  TensorC(1)
```

e.g. modifying fgradhessianv to have:
```fsharp
        let vrev = v |> dsharp.reverseDiff tag
```
in
```fsharp
    static member fgradhessianv f (x:Tensor) (v:Tensor) =
        if x.nelement <> v.nelement then failwithf "x and v must have the same number of elements"
        let tag = GlobalNestingLevel.Next()
        let vrev = v |> dsharp.reverseDiff tag
        let xrev = x |> dsharp.reverseDiff tag
        let fx, gv = dsharp.fgradv f xrev vrev
        gv.reverse()
        fx.primal, gv.primal, xrev.derivative
```

**Comment by gbaydin** [comment]
> This intuitively makes sense to me as fgradv is a function of both x and v and so applying reverse-mode to them both feels like it makes some sense.

This wouldn't be correct as it means that a derivative with respect to v is involved, which is not the case by definition.

**Comment by gbaydin** [comment]
#333 solved this issue and added a test case specifically about this.

I investigated this issue in depth and concluded that this was not a bug. It's a case of taking the second derivative of a function that is not twice differentiable. Getting the exception made sense since there was nothing to propagate for the second-order derivatives. The exception could have been about this specific issue and be more informative.

There are two options to handle such a case. We either throw an exception informing the user or we define the derivative to be all zeros of the correct shape (this is the mathematically correct result but not achievable by autodiff propagation). I noticed PyTorch does have similar behavior with an optional `strict` parameter that selects between these two behaviors.

The fix implements the "define derivative to be all zeros of the correct shape" behavior as the default. This is done at the lowest level of reverse propagation as the most general case. If we choose to implement a strict mode, this would be the place to throw the exception:
https://github.com/DiffSharp/DiffSharp/blob/a670d05a25fa2140303dd40d930e1d9b788f2726/src/DiffSharp.Core/DiffSharp.fs#L1291

**Comment by dsyme** [comment]
Really? Isn't 

    let f (x:Tensor) = x.sum()

twice differentiable?

**Comment by gbaydin** [comment]
Sorry, let me be more precise: the function `.sum()` is indeed twice differentiable in mathematics bu this result is not achievable in an autodiff implementation (except by catching this situation and setting the derivative to all zeros) because `.sum()` is a case where the jacobian of `x.sum()` is independent of the input `x`. You can see how the same situation is handled in PyTorch autograd here: https://github.com/pytorch/pytorch/blob/75024e228ca441290b6a1c2e564300ad507d7af6/torch/autograd/functional.py#L150

**Comment by dsyme** [comment]
Tbh my impression is this is just compensating for the use of scalar zero as an optimised representation (for the zero tensor of appropriate shape) during backprop. And filling in the zeros is thus necessary at edge cases

But either way, the fix is right.
